File: docs/demo/V4.x/TESS_PythonAPI_demo/V4公众号案例代码/行人/MyNet.py
# -*- coding: utf-8 -*-
import collections
import os
import time
from pathlib import Path
import sys
import PySide2
from PySide2.QtCore import *
from PySide2.QtGui import *
from PySide2.QtWidgets import *


from Tessng import TessInterface, TessPlugin, NetInterface, PyCustomerNet
from Tessng import tessngPlugin, tessngIFace, m2p, p2m, Online
from Tessng import NetItemType, GraphicsItemPropName


# 用户插件子类，代表用户自定义与路网相关的实现逻辑，继承自MyCustomerNet
class MyNet(PyCustomerNet):
    def __init__(self):
        super(MyNet, self).__init__()
        iface = tessngIFace()
        netiface = iface.netInterface()
    #
    # # 过载的父类方法，当打开网后TESS NG调用此方法
    # #     实现的逻辑是：路网加载后获取路段数，如果路网数为0则调用方法createNet构建路网，之后再次获取路段数，如果大于0则启动仿真
    def afterLoadNet(self):
        # 代表TESS NG的接口
        iface = tessngIFace()
        # 代表TESS NG的路网子接口
        netiface = iface.netInterface()

        # 创建人行道，将路段向右侧平移offset的距离，得到路侧人行道的形状
        offset = 3
        linkPairs = {"西进口": (1, 6), "南进口": (5, 2), "东进口": (3, 7), "北进口": (8, 4)}
        pathStartPoints = list()  # 记录行人发生点，用于创建行人路径
        pathEndPoints = list()  # 记录行人结束点，用于创建行人路径
        crossWalks = list()  # 记录创建的人行横道，用于在人行横道上创建信号灯
        pedestrianStartPoint = None
        for direction, linkPair in linkPairs.items():
            update_vertexs = []
            for link in linkPair:
                link = netiface.findLink(link)
                vertexs = list()
                rightBreakPoints = link.rightBreakPoints()
                startPoint = rightBreakPoints[0]
                endPoint = rightBreakPoints[-1]
                delta_x = p2m(endPoint.x()) - p2m(startPoint.x())
                delta_y = p2m(endPoint.y()) - p2m(startPoint.y())
                for point in link.rightBreakPoints():
                    x = p2m(point.x())
                    y = p2m(point.y())
                    vertexs.append((x, y))
                if abs(delta_x) > abs(delta_y):
                    if delta_x > 0:
                        update_vertexs += [QPointF(p[0], p[1] + offset) for p in vertexs]
                    else:
                        update_vertexs += [QPointF(p[0], p[1] - offset) for p in vertexs]
                else:
                    if delta_y > 0:
                        update_vertexs += [QPointF(p[0] - offset, p[1]) for p in vertexs]
                    else:
                        update_vertexs += [QPointF(p[0] + offset, p[1]) for p in vertexs]
            new_pedestrianSideWalkRegion = netiface.createPedestrianSideWalkRegion(update_vertexs)
            # 创建行人发生点
            pathStartPoint_x = (update_vertexs[0].x() * 2 + update_vertexs[1].x()) / 3
            pathStartPoint_y = (update_vertexs[0].y() * 2 + update_vertexs[1].y()) / 3
            pathStartPoint = QPointF(pathStartPoint_x, pathStartPoint_y)
            new_pathStartPoint = netiface.createPedestrianPathStartPoint(pathStartPoint)
            pathStartPoints.append(new_pathStartPoint)
            # 创建行人结束点
            pathEndPoint_x = (update_vertexs[-1].x() * 2 + update_vertexs[-2].x()) / 3
            pathEndPoint_y = (update_vertexs[-1].y() * 2 + update_vertexs[-2].y()) / 3
            pathEndPoint = QPointF(pathEndPoint_x, pathEndPoint_y)
            new_pathEndPoint = netiface.createPedestrianPathEndPoint(pathEndPoint)
            pathEndPoints.append(new_pathEndPoint)
            # 记录西进口行人发生点
            if direction == "西进口":
                pedestrianStartPoint = new_pathStartPoint
            print(f"创建{direction}人行道{new_pedestrianSideWalkRegion.getId()},行人发生点{new_pathStartPoint.getId()},行人结束点{new_pathEndPoint.getId()}")
        time.sleep(5)

        # 创建人行横道
        crossWalkPoints = {"西进口": ((-32, -35), (-32, 2)), "北进口": ((-21, 10), (35, 10)),
                           "南进口": ((-24, -39), (33, -39)), "东进口": ((45, -29), (45, 4))}  # 人行横道起终点坐标
        pathMiddlePoints = list()  # 记录人行横道中心点，用于创建行人路径
        for direction, crossWalkPoint in crossWalkPoints.items():
            startPointCoor = crossWalkPoint[0]
            startPoint = QPointF(startPointCoor[0], -startPointCoor[1])
            endPointCoor = crossWalkPoint[1]
            endPoint = QPointF(endPointCoor[0], -endPointCoor[1])
            # 取人行横道中心点，用于创建行人路径
            pathMiddlePoint_x = (startPointCoor[0] + endPointCoor[0]) / 2
            pathMiddlePoint_y = - (startPointCoor[1] + endPointCoor[1]) / 2
            pathMiddlePoint = QPointF(pathMiddlePoint_x, pathMiddlePoint_y)
            crossWalk = netiface.createPedestrianCrossWalkRegion(startPoint, endPoint)
            crossWalk.setName(f"{direction}人行横道")
            crossWalks.append(crossWalk)
            pathMiddlePoints.append(pathMiddlePoint)
            print(f"创建{direction}人行横道{crossWalk.getId()}")
        time.sleep(5)

        # 创建过街行人路径
        straightPath = netiface.createPedestrianPath(pathStartPoints[0], pathEndPoints[1], [pathMiddlePoints[2]])  # 西进口直行过街
        print("创建西进口直行过街路径")
        leftPath = netiface.createPedestrianPath(pathStartPoints[0], pathEndPoints[3], [pathMiddlePoints[0]])  # 西进口左转过街
        print("创建西进口左转过街路径")
        netiface.createPedestrianPath(pathStartPoints[1], pathEndPoints[2], [pathMiddlePoints[3]])  # 南进口直行过街
        print("创建南进口直行过街路径")
        netiface.createPedestrianPath(pathStartPoints[2], pathEndPoints[3], [pathMiddlePoints[1]])  # 东进口直行过街
        print("创建东进口直行过街路径")
        netiface.createPedestrianPath(pathStartPoints[3], pathEndPoints[0], [pathMiddlePoints[0]])  # 北进口直行过街
        print("创建北进口直行过街路径")
        time.sleep(5)

        # # 创建人行横道信号灯
        crossWalkPhaseId = {"西进口": 5, "东进口": 5, "南进口": 4, "北进口": 4}  # 东西进口行人采用南北方向机动车相位，南北进口行人采用东西方向机动车相位
        trafficController = netiface.findSignalControllerById(1)
        # 遍历人行横道创建信号灯
        for crossWalk in crossWalks:
            crossWalkId = crossWalk.getId()
            crossWalkName = crossWalk.getName()
            crossWalkDirection = crossWalkName.replace('人行横道', '')
            crossWalkPoint = crossWalkPoints.get(crossWalkDirection)
            crossWalkStartPoint = crossWalkPoint[0]
            crossWalkEndPoint = crossWalkPoint[1]
            name = f"{crossWalkName}信号灯"
            isPositive = True
            position_x = (crossWalkStartPoint[0] * 9 + crossWalkEndPoint[0]) / 10
            position_y = (crossWalkStartPoint[1] * 9 + crossWalkEndPoint[1]) / 10
            position = QPointF(position_x, -position_y)
            signalLamp_positive = netiface.createCrossWalkSignalLamp(trafficController, name, crossWalkId, position,
                                                                     isPositive)
            isPositive = False
            position_x = (crossWalkStartPoint[0] + crossWalkEndPoint[0] * 9) / 10
            position_y = (crossWalkStartPoint[1] + crossWalkEndPoint[1] * 9) / 10
            position = QPointF(position_x, -position_y)
            signalLamp_negtive = netiface.createCrossWalkSignalLamp(trafficController, name, crossWalkId, position,
                                                                    isPositive)

            # 为信号灯分配相位
            phase = netiface.findSignalPhase(crossWalkPhaseId.get(crossWalkDirection))
            netiface.addCrossWalkSignalPhaseToLamp(phase.id(), signalLamp_positive)
            netiface.addCrossWalkSignalPhaseToLamp(phase.id(), signalLamp_negtive)
            print(f"为{crossWalkDirection}人行横道创建双向信号灯{signalLamp_positive.id()},{signalLamp_negtive}; 并分配相位{phase.id()}")

        # 更新行人发生点流量和路径分配比例
        configInfo = netiface.findPedestrianStartPointConfigInfo(pedestrianStartPoint.getId())
        update_configInfo = Online.Pedestrian.PedestrianPathStartPointConfigInfo()
        # # 更新行人发生点流量
        update_genPedestrianConfigInfo = Online.Pedestrian.GenPedestrianInfo()
        update_genPedestrianConfigInfo.pedestrianCount = 6
        update_genPedestrianConfigInfo.timeInterval = configInfo.genPedestrianConfigInfo[0].timeInterval
        update_configInfo.genPedestrianConfigInfo = [update_genPedestrianConfigInfo]
        update_configInfo.id = configInfo.id
        distributionInfo = configInfo.pedestrianTrafficDistributionConfigInfo[0]
        update_distributionInfo = Online.Pedestrian.PedestrianTrafficDistributionInfo()
        update_distributionInfo.timeInterval = distributionInfo.timeInterval
        update_distributionInfo.trafficRatio = {straightPath.getId(): 2, leftPath.getId(): 1}
        update_configInfo.pedestrianTrafficDistributionConfigInfo = [update_distributionInfo]
        result = netiface.updatePedestrianStartPointConfigInfo(update_configInfo)
        print(f"更新西进口行人发生点流量:{update_genPedestrianConfigInfo.pedestrianCount}人/{update_genPedestrianConfigInfo.timeInterval}秒,result:{result}")
        print(f"更新西进口行人发生点路径分配:{update_distributionInfo.trafficRatio},result:{result}")

File: docs/demo/V4.x/TESS_CppAPI_DEMO/1.基础功能/行人仿真/MyNet.cpp
#include "MyNet.h"
#include <QList>
#include <QMap>
#include <QDebug>
#include "tessinterface.h"
#include "netinterface.h"
#include "UnitChange.h"
#include "IPedestrianRectRegion.h"
#include "IPedestrianRegion.h"
#include "IPedestrianPathPoint.h"
#include "IPedestrianPath.h"
#include "IPedestrianSideWalkRegion.h"
#include "IPedestrianCrossWalkRegion.h"
#include "ISignalPhase.h"
#include "ilink.h"
#include "ICrosswalkSignalLamp.h"

MyNet::MyNet(){
}

//加载完路网后的行为
void MyNet::afterLoadNet() {
    setupPedestrianDemo();
}

MyNet::~MyNet(){
}

void MyNet::setupPedestrianDemo()
{
    auto ni = gpTessInterface->netInterface();
    qreal offset = 3;
    QMap<QString, QPair<long, long>> linkPairs;
    linkPairs.insert("西进口", QPair<long, long>(1, 6));
    linkPairs.insert("南进口", QPair<long, long>(5, 2));
    linkPairs.insert("东进口", QPair<long, long>(3, 7));
    linkPairs.insert("北进口", QPair<long, long>(8, 4));

    QList<IPedestrianPathPoint*> pathStartPoints;
    QList<IPedestrianPathPoint*> pathEndPoints;
    QList<IPedestrianCrossWalkRegion*> crossWalks;
    IPedestrianPathPoint* pedestrianStartPoint = nullptr;

    for (auto it = linkPairs.constBegin(); it != linkPairs.constEnd(); ++it) {
        QString direction = it.key();
        QPair<long, long> pair = it.value();
        QList<QPointF> update_vertexs;
        QList<long> ids; ids << pair.first << pair.second;
        for (long linkId : ids) {
            ILink* link = ni->findLink(linkId);
            if (!link) continue;
            QList<QPointF> rightPts = link->rightBreakPoints();
            if (rightPts.isEmpty()) continue;
            QPointF startPoint = rightPts.first();
            QPointF endPoint = rightPts.last();
            qreal delta_x = p2m(endPoint.x()) - p2m(startPoint.x());
            qreal delta_y = p2m(endPoint.y()) - p2m(startPoint.y());
            QList<QPointF> vertexs;
            for (const QPointF& pt : rightPts) {
                vertexs.append(QPointF(p2m(pt.x()), p2m(pt.y())));
            }
            if (qAbs(delta_x) > qAbs(delta_y)) {
                if (delta_x > 0) {
                    for (const QPointF& p : vertexs) update_vertexs.append(QPointF(p.x(), p.y() + offset));
                } else {
                    for (const QPointF& p : vertexs) update_vertexs.append(QPointF(p.x(), p.y() - offset));
                }
            } else {
                if (delta_y > 0) {
                    for (const QPointF& p : vertexs) update_vertexs.append(QPointF(p.x() - offset, p.y()));
                } else {
                    for (const QPointF& p : vertexs) update_vertexs.append(QPointF(p.x() + offset, p.y()));
                }
            }
        }
        if (update_vertexs.size() < 2) {
            qDebug() << "方向" << direction << "人行道顶点不足，跳过创建，点数:" << update_vertexs.size();
            continue;
        }
        IPedestrianSideWalkRegion* sideWalk = ni->createPedestrianSideWalkRegion(update_vertexs);
        QPointF pathStartPoint((update_vertexs[0].x() * 2 + update_vertexs[1].x()) / 3.0,
                               (update_vertexs[0].y() * 2 + update_vertexs[1].y()) / 3.0);
        IPedestrianPathPoint* newStart = ni->createPedestrianPathStartPoint(pathStartPoint);
        if (newStart) pathStartPoints.append(newStart);
        QPointF pathEndPoint((update_vertexs.last().x() * 2 + update_vertexs[update_vertexs.size() - 2].x()) / 3.0,
                             (update_vertexs.last().y() * 2 + update_vertexs[update_vertexs.size() - 2].y()) / 3.0);
        IPedestrianPathPoint* newEnd = ni->createPedestrianPathEndPoint(pathEndPoint);
        if (newEnd) pathEndPoints.append(newEnd);
        if (direction == "西进口") {
            pedestrianStartPoint = newStart;
        }
        qDebug() << "创建" << direction << "人行道" << (sideWalk ? sideWalk->getId() : -1)
                 << ",行人发生点" << (newStart ? newStart->getId() : -1)
                 << ",行人结束点" << (newEnd ? newEnd->getId() : -1);
    }

    QMap<QString, QPair<QPointF, QPointF>> crossWalkPoints;
    crossWalkPoints.insert("西进口", QPair<QPointF, QPointF>(QPointF(-32, -35), QPointF(-32, 2)));
    crossWalkPoints.insert("北进口", QPair<QPointF, QPointF>(QPointF(-21, 10), QPointF(35, 10)));
    crossWalkPoints.insert("南进口", QPair<QPointF, QPointF>(QPointF(-24, -39), QPointF(33, -39)));
    crossWalkPoints.insert("东进口", QPair<QPointF, QPointF>(QPointF(45, -29), QPointF(45, 4)));
    QList<QPointF> pathMiddlePoints;
    for (auto it = crossWalkPoints.constBegin(); it != crossWalkPoints.constEnd(); ++it) {
        QString direction = it.key();
        QPointF startPointCoor = it.value().first;
        QPointF endPointCoor = it.value().second;
        QPointF startPoint(startPointCoor.x(), -startPointCoor.y());
        QPointF endPoint(endPointCoor.x(), -endPointCoor.y());
        QPointF mid((startPointCoor.x() + endPointCoor.x()) / 2.0,
                    -((startPointCoor.y() + endPointCoor.y()) / 2.0));
        IPedestrianCrossWalkRegion* crossWalk = ni->createPedestrianCrossWalkRegion(startPoint, endPoint);
        if (crossWalk) crossWalk->setName(QString("%1人行横道").arg(direction));
        crossWalks.append(crossWalk);
        pathMiddlePoints.append(mid);
        qDebug() << "创建" << direction << "人行横道" << (crossWalk ? crossWalk->getId() : -1);
    }

    IPedestrianPath* straightPath = nullptr;
    IPedestrianPath* leftPath = nullptr;
    if (pathStartPoints.size() >= 4 && pathEndPoints.size() >= 4 && pathMiddlePoints.size() >= 4) {
        straightPath = ni->createPedestrianPath(pathStartPoints[0], pathEndPoints[1], QList<QPointF>() << pathMiddlePoints[2]);
        qDebug() << "创建西进口直行过街路径";
        leftPath = ni->createPedestrianPath(pathStartPoints[0], pathEndPoints[3], QList<QPointF>() << pathMiddlePoints[0]);
        qDebug() << "创建西进口左转过街路径";
        ni->createPedestrianPath(pathStartPoints[1], pathEndPoints[2], QList<QPointF>() << pathMiddlePoints[3]);
        qDebug() << "创建南进口直行过街路径";
        ni->createPedestrianPath(pathStartPoints[2], pathEndPoints[3], QList<QPointF>() << pathMiddlePoints[1]);
        qDebug() << "创建东进口直行过街路径";
        ni->createPedestrianPath(pathStartPoints[3], pathEndPoints[0], QList<QPointF>() << pathMiddlePoints[0]);
        qDebug() << "创建北进口直行过街路径";
    } else {
        qDebug() << "路径创建失败，起止点或中间点数量不足: start" << pathStartPoints.size() << "end" << pathEndPoints.size() << "middle" << pathMiddlePoints.size();
    }

    QMap<QString, int> crossWalkPhaseId; crossWalkPhaseId.insert("西进口", 5); crossWalkPhaseId.insert("东进口", 5); crossWalkPhaseId.insert("南进口", 4); crossWalkPhaseId.insert("北进口", 4);
    ISignalController* trafficController = ni->findSignalControllerById(1);
    for (IPedestrianCrossWalkRegion* crossWalk : crossWalks) {
        if (!crossWalk) continue;
        long crossWalkId = crossWalk->getId();
        QString crossWalkName = crossWalk->getName();
        QString crossWalkDirection = crossWalkName;
        crossWalkDirection.replace("人行横道", "");
        auto pair = crossWalkPoints.value(crossWalkDirection);
        QPointF s = pair.first; QPointF e = pair.second;
        QString name = QString("%1信号灯").arg(crossWalkName);
        bool isPositive = true;
        QPointF pos1((s.x() * 9 + e.x()) / 10.0, -((s.y() * 9 + e.y()) / 10.0));
        auto lampPos = ni->createCrossWalkSignalLamp(trafficController, name, crossWalkId, pos1, isPositive);
        isPositive = false;
        QPointF pos2((s.x() + e.x() * 9) / 10.0, -((s.y() + e.y() * 9) / 10.0));
        auto lampNeg = ni->createCrossWalkSignalLamp(trafficController, name, crossWalkId, pos2, isPositive);
        ISignalPhase* phase = ni->findSignalPhase(crossWalkPhaseId.value(crossWalkDirection));
        if (phase) {
            ni->addCrossWalkSignalPhaseToLamp(phase->id(), lampPos);
            ni->addCrossWalkSignalPhaseToLamp(phase->id(), lampNeg);
        }
        qDebug() << "为" << crossWalkDirection << "人行横道创建双向信号灯" << (lampPos ? lampPos->id() : -1) << "," << (lampNeg ? lampNeg->id() : -1) << "; 并分配相位" << (phase ? phase->id() : -1);
    }

    if (pedestrianStartPoint && straightPath && leftPath) {
        auto configInfo = ni->findPedestrianStartPointConfigInfo(pedestrianStartPoint->getId());
        Online::Pedestrian::PedestrianPathStartPointConfigInfo updateConfig;
        Online::Pedestrian::GenPedestrianInfo gen;
        gen.pedestrianCount = 6;
        gen.timeInterval = configInfo.genPedestrianConfigInfo.isEmpty() ? 600 : configInfo.genPedestrianConfigInfo.first().timeInterval;
        updateConfig.genPedestrianConfigInfo = QList<Online::Pedestrian::GenPedestrianInfo>() << gen;
        updateConfig.id = configInfo.id;
        Online::Pedestrian::PedestrianTrafficDistributionInfo dis;
        dis.timeInterval = configInfo.pedestrianTrafficDistributionConfigInfo.isEmpty() ? 600 : configInfo.pedestrianTrafficDistributionConfigInfo.first().timeInterval;
        dis.trafficRatio.insert((int)(straightPath ? straightPath->getId() : -1), 2);
        dis.trafficRatio.insert((int)(leftPath ? leftPath->getId() : -1), 1);
        updateConfig.pedestrianTrafficDistributionConfigInfo = QList<Online::Pedestrian::PedestrianTrafficDistributionInfo>() << dis;
        bool result = ni->updatePedestrianStartPointConfigInfo(updateConfig);
        qDebug() << "更新西进口行人发生点流量:" << gen.pedestrianCount << "人/" << gen.timeInterval << "秒,result:" << result;
        qDebug() << "更新西进口行人发生点路径分配:" << dis.trafficRatio << ",result:" << result;
    }
}